### ISSUE TITLE
RFC: lock readouts behind their respective features

### ISSUE THREAD

**Issue by grtcdr** [Macchina-CLI/libmacchina#110]
I think libmacchina can greatly benefit if we were to lock each readout type
behind its respective feature.
 
I believe this would entice developers that are not willing to ship their
programs with a library that brings a handful of dependencies, and I am one of
them (sort of). As I've been busy working on citron, which uses libmacchina as its
backend, I've had to deal with dependencies that just don't make any sense for
a program like citron.

What are your thoughts?

**Comment by grtcdr** [comment]
@uttarayan21 mentioned in our chat in the (lib)macchina matrix room, that we continue to package libmacchina with all features enabled by default (if we were to implement this proposed change).

In other words, application developers that are interested in libmacchina can either use all the readouts we offer (by default), or cherry-pick the readouts they want to use.

I totally agree with this, this is our best bet to keep things simple on the outside.

**Comment by grtcdr** [comment]
Here are some of the new readouts I propose:
```rust
pub trait GraphicalReadout {
    /// Creates a new instance of the structure.
    fn new() -> Self;
    /// TODO: documentation
    fn desktop_environment(&self) -> Result<String, ReadoutError>;
    /// TODO: documentation
    fn session(&self) -> Result<String, ReadoutError>;
    /// TODO: documentation
    fn window_manager(&self) -> Result<String, ReadoutError>;
    /// TODO: documentation
    fn terminal(&self) -> Result<String, ReadoutError>;
    /// TODO: documentation
    fn shell(&self, format: ShellFormat, kind: ShellKind) -> Result<String, ReadoutError>;
}
```
```rust
pub trait ProcessorReadout {
    /// Creates a new instance of the structure.
    fn new() -> Self;
    /// TODO: documentation
    fn cpu_model_name(&self) -> Result<String, ReadoutError>;
    /// TODO: documentation
    fn cpu_usage(&self) -> Result<usize, ReadoutError>;
    /// TODO: documentation
    fn cpu_physical_cores(&self) -> Result<usize, ReadoutError>;
    /// TODO: documentation
    fn cpu_cores(&self) -> Result<usize, ReadoutError>;
}
```

**Comment by grtcdr** [comment]
The linked pull request was written purely as a demonstration, it's only working on Linux right now, since it makes some pretty bold breaking changes to the API, but it does deliver on the one thing that it promises: modularity.

**Comment by grtcdr** [comment]
https://user-images.githubusercontent.com/35816711/147727323-958fb622-cd83-4709-a795-5d5151d952a1.mp4

**Comment by FantasyTeddy** [comment]
I really like the idea of allowing users of the library choose which parts (dependencies) are needed for their application.

However there is one concern that I have with introducing even more readouts: How can we share state between multiple different readouts?
Let me try to explain:
- **Example 1:** The Linux implementation of multiple readouts stores a `sysinfo` struct and even though this is not a particularly expensive operation, we could imagine that if we wanted to speed things up, we would maybe also perform the call to `sysinfo()` only once. Therefore it would be desirable to be able to store such state information at one central place instead of once per readout.
- **Example 2:** As mentioned in #112 the WMI information on Windows will probably be used even more when implementing more readouts (like resolution or backlight). For querying the WMI information a connection needs to be opened which is quite expensive. It would therefore also be beneficial if the connection could be opened once and reused for each separate query.

This concern is only slightly related to this proposed change, but I wanted to share my thoughts on this and maybe you have some ideas how we could solve this (if we need to).

**Comment by grtcdr** [comment]
I think it's important that you brought this up, I have always wondered how many milliseconds we can shave off just by saving a state instead of reinstantiating the necessary structs.

I haven't experimented with a solution, since the Linux implementation is so fast, any gained time is barely noticeable, but I always like seeing smaller numbers. Also, I'm unhappy with the Windows performance, so we definitely need to do something about this.

I can't say I know how to approach this correctly and idiomatically, but I will experiment with some ideas. We should have a separate issue for this, because we need to discuss how to translate this idea into code properly.

Before we continue with this, I want to see what the others think of this RFC, so we can all pour our best work in #111 instead of main.

**Comment by grtcdr** [comment]
I believe it won't be possible to share objects between traits without compromising on the "one trait multiple implementations" setup that we've got.

I think that if we truly desire the utmost performance, most if not all platforms will have to have their own implementations without a trait to adhere by, very similar to how libc works.

**Comment by FantasyTeddy** [comment]
One way to share some state would be to go the other way: reduce the number of different readouts to only one and every implementation can define their own shared state.
However I don't think that this is necessarily a good solution, since I like the way that the current readouts are split up to make it clear what information they offer.
I also don't know how (and if) this would work with the proposed feature flags.

**Comment by grtcdr** [comment]
> One way to share some state would be to go the other way: reduce the number of different readouts to only one and every implementation can define their own shared state.

Well, there wouldn't be much of a point to this RFC if we were to do this, unfortunately. And like you said, we would lose out on _context_.

**Comment by grtcdr** [comment]
> I also don't know how (and if) this would work with the proposed feature flags.

I almost managed to get this to work, until I hit a roadblock I'm not sure I can recover from. Here's what I've been staring at for the last 40 minutes:

```rust
calls in constants are limited to constant functions, tuple structs and tuple variants
  --> src/linux/mod.rs:49:46
   |
49 |         const return_value: c_int = unsafe { sysinfo(sys_ptr) };
```

**Comment by grtcdr** [comment]
I'll try and get some benchmarks, but I don't want to get your hopes up, the only benefit of this is less allocations (which is still good, to be honest). libmacchina makes way too many of them, just take a look at `valgrind macchina`.

**Comment by FantasyTeddy** [comment]
> Well, there wouldn't be much of a point to this RFC if we were to do this, unfortunately. And like you said, we would lose out on _context_.

I'm not very familiar how features work in Rust, but I imagined that you could also exclude/include single functions instead of entire structs. If this is not how this works then this RFC would indeed be incompatible with my suggestion.

> I'll try and get some benchmarks, but I don't want to get your hopes up, the only benefit of this is less allocations (which is still good, to be honest). libmacchina makes way too many of them, just take a look at `valgrind macchina`.

Even though it would certainly be better to do the allocation only once, I agree that there is not much performance to be gained here. I think the second example I mentioned before (opening a WMI connection on Windows) would be a better argument where performance matters, since this is quite an expensive operation.

**Comment by grtcdr** [comment]
> Even though it would certainly be better to do the allocation only once, I agree that there is not much performance to be gained here. I think the second example I mentioned before (opening a WMI connection on Windows) would be a better argument where performance matters, since this is quite an expensive operation.

The only possible way to achieve this as far as I can tell (I've been trying to piece it all together since you first suggested this idea) is to drop our traits module and have every platform be its own thing. This way, you could start the connection in the constructor, and pass a reference to the connection to all functions that need it.

**Comment by FantasyTeddy** [comment]
Exactly.
I guess you got my point and we can leave it at that for the moment, because like you just mentioned (and already mentioned before) this would probably require an entirely different discussion and is kind of off-topic here.

**Comment by grtcdr** [comment]
I've already reverted the changes I previously pushed, the complexity that it introduced is just not worth it, hopefully we can pick this up some other day.

**Comment by grtcdr** [comment]
> I'm not very familiar how features work in Rust

It's my first time making use of features, so I'm no expert either.

> but I imagined that you could also exclude/include single functions instead of entire structs

That is totally possible, just about anything can be made a feature, even single lines of code.

**Comment by rashil2000** [comment]
> I think the second example I mentioned before (opening a WMI connection on Windows) would be a better argument where performance matters, since this is quite an expensive operation.

As @FantasyTeddy already mentions, a lot readouts will require WMI queries. (In winfetch too we use it for around 10-12 readouts, but all that is done through a single connection - called `CimSession` shown [here](https://github.com/kiedtl/winfetch/blob/d14361e3d2f2e29e7a7f4f30e0dcd33bfbe135b4/winfetch.ps1#L175)). 

Will it be possible to lock readouts that require WMI behind one feature? Instead of having separate features for example for OSName, Backlight and Resolution, we can one feature for stuff fetched through WMI. 

On a similar note, will it be possible to bunch-up all Linux readouts together that require the `sysinfo` dependency? I assume it's being used to provide a multitude of information categories.

**Comment by grtcdr** [comment]
Your request contradicts what the RFC proposes.

If we were to implement shared objects (or a connection for Windows), it MUST be done at the root level of the platform e.g. `windows/mod.rs` or `linux/mod.rs` and the connection should only be opened when a feature that requires it is enabled.

**Comment by rashil2000** [comment]
That makes sense, thanks for the info.
